java/tsfile/src/main/java/org/apache/tsfile/read/filter/operator/ValueFilterOperators.java

```diff
@@ -765,7 +771,89 @@ public ValueIn(ByteBuffer buffer) {
 
     @Override
     public boolean valueSatisfy(Object value) {
-      return candidates.contains(value);
+      return candidates.contains((T) value);
+    }
+
+    @Override
+    public boolean canSkip(IMetadata metadata) {
+      Optional<Statistics<? extends Serializable>> statistics =
+          metadata.getMeasurementStatistics(measurementIndex);
+
+      // All values are null, but candidates do not contain null
+      if ((!statistics.isPresent() || isAllNulls(statistics.get())) && !candidates.contains(null)) {
+        return true;
+      }
+
+      // All values are not null, but candidate is one null value
+      if (metadata.hasNullValue(measurementIndex)
+          && candidates.size() == 1
+          && candidates.contains(null)) {
+        return true;
+      }
+
+      if (statistics.isPresent()) {
+        T valuesMin = (T) statistics.get().getMinValue();
+        T valuesMax = (T) statistics.get().getMaxValue();
+        // All values are same
+        if (valuesMin == valuesMax) {
+          for (T val : candidates) {
+            if (valuesMin == val) {
+              return false;
+            }
+          }
+          return true;
+        } else {
+          // All values are less than min, or greater than max
+          if (candidatesMin.compareTo(valuesMax) > 0) {
+            return true;
+          }
+          if (candidatesMax.compareTo(valuesMax) < 0) {
+            return true;
+          }
+        }
+      }
+
+      return false;
+    }
+
+    @Override
+    protected boolean canSkip(Statistics<? extends Serializable> statistics) {
+      throw new NotImplementedException();
+    }
+
+    @Override
+    public boolean allSatisfy(IMetadata metadata) {
+      Optional<Statistics<? extends Serializable>> statistics =
+          metadata.getMeasurementStatistics(measurementIndex);
+
+      // All values are null, and candidate is one null
+      if ((!statistics.isPresent() || isAllNulls(statistics.get()))
+          && candidates.size() == 1
+          && candidates.contains(null)) {
+        return true;
+      }
+
+      // All values are same
+      if (statistics.isPresent()) {
+        T valuesMin = (T) statistics.get().getMinValue();
+        T valuesMax = (T) statistics.get().getMaxValue();
+        // All values are same
+        if (valuesMin == valuesMax) {
+          for (T val : candidates) {
+            if (valuesMin != val) {
+              return false;
```

Comment:
wrong?

Any values in set equals to valuesMin, should return true

Reply:
Fixed.